src/cryptsetup/cryptsetup.c

```diff
@@ -497,16 +496,7 @@ static int attach_luks_or_plain(struct crypt_device *cd,
         assert(name);
         assert(key_file || passwords);
 
-        if (!arg_type || STR_IN_SET(arg_type, ANY_LUKS, CRYPT_LUKS1)) {
-                r = crypt_load(cd, CRYPT_LUKS, NULL);
-                if (r < 0)
-                        return log_error_errno(r, "Failed to load LUKS superblock on device %s: %m", crypt_get_device_name(cd));
-
-                if (data_device)
-                        r = crypt_set_data_device(cd, data_device);
-        }
-
-        if ((!arg_type && r < 0) || streq_ptr(arg_type, CRYPT_PLAIN)) {
+        if ((!arg_type && !crypt_get_type(cd)) || streq_ptr(arg_type, CRYPT_PLAIN)) {
```

Comment:
hmm, this changes behaviour, no? previously, if crypt_set_data_device() failed we'd try this section, while now we'll fail fatally.

Reply:
Yes, and I think that to continue if data device load fails is a bug (it is mentioned in the patch header). If data device load fails, you should not try to map it as plain device - it have LUKS superblock (so we already know it is LUKS device), just something is wrong with data (perhaps wrong crypttab?) IMO it can lead to data corruption (using different key in plain mode).

Reply:
I have suspicion that some people are relying on such an automatism... i mean, i am not against changing this behaviour, but this should not be done as part of other work, but in an explicit commit of its own, any maybe even come with a NEWS change declaring that we don't fall back from LUKS to plain anymore if we can't find LUKS. Because that's really what this comes down to, no?